File: bandhu_app/bandhu_app/page/nurse_form/nurse_form.py
import frappe
from frappe import _
from frappe.query_builder.functions import Count
from frappe.utils import flt

from bandhu_app.bandhu_app.utils.patient_details import get_patient_details, get_session_encounters
from bandhu_app.bandhu_app.utils.realtime import publish_board_update
from bandhu_app.bandhu_app.utils.session import find_active_session, find_upcoming_sessions


def require_session_access(session_name: str) -> None:
	user = frappe.session.user
	roles = frappe.get_roles(user)
	if "System Manager" in roles:
		return
	if "Nurse" not in roles:
		frappe.throw(
			_("You do not have permission to access this clinic session."),
			frappe.PermissionError,
		)
	practitioner = frappe.db.get_value("Healthcare Practitioner", {"user_id": user}, "name")
	if not practitioner:
		frappe.throw(
			_("No Healthcare Practitioner linked to your account."),
			frappe.PermissionError,
		)
	assigned_nurse = frappe.db.get_value("Bandhu Clinic Session", session_name, "assigned_nurse")
	if not assigned_nurse or assigned_nurse != practitioner:
		frappe.throw(
			_("You are not assigned to this clinic session."),
			frappe.PermissionError,
		)


def get_nurse_practitioner():
	return frappe.db.get_value("Healthcare Practitioner", {"user_id": frappe.session.user}, "name")


def load_session_encounter(encounter: str):
	doc = frappe.get_doc("Patient Encounter", encounter)
	require_session_access(doc.custom_clinic_session)
	return doc


@frappe.whitelist()
def get_session_status() -> dict:
	user = frappe.session.user

	roles = frappe.get_roles(user)
	if "Nurse" not in roles and "System Manager" not in roles:
		return {"has_session": False, "message": _("You do not have the Nurse role.")}

	practitioner = frappe.db.get_value(
		"Healthcare Practitioner",
		{"user_id": user},
		"name",
	)

	if not practitioner:
		return {"has_session": False, "message": _("No Healthcare Practitioner linked to your account.")}

	session = find_active_session("assigned_nurse", practitioner)

	if not session:
		return {
			"has_session": False,
			"message": _("No session scheduled for today. Please contact Programme Manager."),
		}

	return {
		"has_session": True,
		"session_name": session.name,
		"status": session.status,
		"clinic": session.clinic,
		"site": session.site,
	}


@frappe.whitelist()
def get_upcoming_sessions() -> list:
	roles = frappe.get_roles()
	if "Nurse" not in roles and "System Manager" not in roles:
		frappe.throw(
			_("You do not have permission to access this page."),
			frappe.PermissionError,
		)

	practitioner = get_nurse_practitioner()
	if not practitioner:
		return []
	return find_upcoming_sessions("assigned_nurse", practitioner)


def load_session_for_status_change(session_name: str) -> dict:
	require_session_access(session_name)
	# for_update makes a second open or close request wait and read the committed status.
	session_doc = frappe.db.get_value(
		"Bandhu Clinic Session",
		session_name,
		["status", "date"],
		as_dict=True,
		for_update=True,
	)
	if not session_doc:
		frappe.throw(_("Clinic session not found."))
	if session_doc.status == "Cancelled":
		frappe.throw(_("This session was cancelled. Do not travel to it."))

	return session_doc


@frappe.whitelist(methods=["POST"])
def start_session(session_name: str) -> None:
	session_doc = load_session_for_status_change(session_name)

	if session_doc.status == "In Progress":
		frappe.throw(_("This session is already open."))
	if session_doc.status == "Completed":
		frappe.throw(
			_("This session is already closed and cannot be reopened."),
		)
	if str(session_doc.date) != frappe.utils.today():
		frappe.throw(_("You can only open a session on the day it is scheduled."))

	frappe.db.set_value(
		"Bandhu Clinic Session",
		session_name,
		{"status": "In Progress", "start_time": frappe.utils.now_datetime()},
	)
	publish_board_update(session_name)


@frappe.whitelist(methods=["POST"])
def end_session(session_name: str) -> None:
	session_doc = load_session_for_status_change(session_name)

	if session_doc.status != "In Progress":
		frappe.throw(_("This session is not open, so it cannot be closed."))

	frappe.db.set_value(
		"Bandhu Clinic Session",
		session_name,
		{"status": "Completed", "end_time": frappe.utils.now_datetime()},
	)
	publish_board_update(session_name)


@frappe.whitelist()
def get_patients_for_tests(session_name: str) -> list:
	require_session_access(session_name)
	return get_session_encounters(session_name, "Awaiting Test")


@frappe.whitelist()
def get_patients_for_medicines(session_name: str) -> list:
	require_session_access(session_name)
	return get_session_encounters(session_name, "Awaiting Medicine")


@frappe.whitelist()
def get_completed_patients(session_name: str) -> list:
	require_session_access(session_name)
	return get_session_encounters(session_name, "Completed")


SESSION_PROGRESS_STATES = {
	"registered": "Waiting for Doctor",
	"with_doctor": "Awaiting Doctor Review",
	"for_tests": "Awaiting Test",
	"for_medicines": "Awaiting Medicine",
	"completed": "Completed",
}


@frappe.whitelist()
def get_session_progress(session_name: str) -> dict:
	require_session_access(session_name)

	# v16 rejects an aggregate string in `fields`, so use the query builder.
	encounter = frappe.qb.DocType("Patient Encounter")
	counts = (
		frappe.qb.from_(encounter)
		.select(encounter.custom_workflow_state, Count("*").as_("total"))
		.where(encounter.custom_clinic_session == session_name)
		.groupby(encounter.custom_workflow_state)
	).run(as_dict=True)
	by_state = {row.custom_workflow_state: row.total for row in counts}

	return {key: by_state.get(state, 0) for key, state in SESSION_PROGRESS_STATES.items()}


@frappe.whitelist()
def get_patient_registration_details(encounter: str) -> dict:
	doc = load_session_encounter(encounter)
	return get_patient_details(doc.patient)


@frappe.whitelist(methods=["POST"])
def submit_test_results(encounter: str, results: list | str) -> None:
	doc = load_session_encounter(encounter)
	if doc.custom_workflow_state != "Awaiting Test":
		frappe.throw(_("This patient is not awaiting a test."))

	results = frappe.parse_json(results)
	rows_by_name = {row.name: row for row in doc.custom_test_instructions}
	for result in results:
		row = rows_by_name.get(result.get("name"))
		if not row:
			frappe.throw(_("Unknown test row."))
		row.result_type = result.get("result_type")
		row.result_value = result.get("result_value")

	for row in doc.custom_test_instructions:
		if not row.result_type:
			frappe.throw(
				_("{0} has no result. Choose Not Done if the test could not be run.").format(row.test_name)
			)
		if row.result_type == "Value" and not (row.result_value or "").strip():
			frappe.throw(_("{0} is a value test and needs a reading.").format(row.test_name))

	doc.custom_workflow_state = "Awaiting Doctor Review"
	doc.save(ignore_permissions=True)


@frappe.whitelist(methods=["POST"])
def record_vitals(
	encounter: str,
	height_cm: float | None = None,
	weight_kg: float | None = None,
	temperature: float | None = None,
	pulse_rate: int | None = None,
	spo2: int | None = None,
	bp_systolic: int | None = None,
	bp_diastolic: int | None = None,
) -> None:
	doc = load_session_encounter(encounter)
	if doc.custom_workflow_state not in ("Awaiting Test", "Awaiting Medicine"):
		frappe.throw(_("Vitals can only be recorded while the patient is with the nurse."))

	measurements = [
		(_("Height"), height_cm, 30, 250),
		(_("Weight"), weight_kg, 1, 300),
		(_("Temperature"), temperature, 90, 110),
		(_("Pulse"), pulse_rate, 20, 250),
		(_("SpO2"), spo2, 50, 100),
		(_("BP systolic"), bp_systolic, 50, 300),
		(_("BP diastolic"), bp_diastolic, 30, 200),
	]
	if not any(value is not None for _label, value, _low, _high in measurements):
		frappe.throw(_("Enter at least one vital sign."))

	for label, value, low, high in measurements:
		if value is None:
			continue
		if flt(value) <= 0:
			frappe.throw(_("Vital signs must be positive numbers."))
		if not low <= flt(value) <= high:
			frappe.throw(
				_("{0} of {1} is outside what a person can record. Check the entry.").format(label, value)
			)

	if (bp_systolic is None) != (bp_diastolic is None):
		frappe.throw(_("Blood pressure needs both the systolic and the diastolic number."))
	if bp_systolic is not None and flt(bp_diastolic) >= flt(bp_systolic):
		frappe.throw(_("The systolic number has to be the higher of the two."))

	if height_cm is not None:
		doc.custom_height = height_cm
	if weight_kg is not None:
		doc.custom_weight = weight_kg
	if temperature is not None:
		doc.custom_temperature = temperature
	if pulse_rate is not None:
		doc.custom_pulse_rate = pulse_rate
	if spo2 is not None:
		doc.custom_spo2 = spo2
	if bp_systolic is not None and bp_diastolic is not None:
		doc.custom_blood_pressure = f"{bp_systolic}/{bp_diastolic}"

	if doc.custom_height and doc.custom_weight:
		height_m = flt(doc.custom_height) / 100
		doc.custom_bmi = round(flt(doc.custom_weight) / (height_m * height_m), 2)

	doc.save(ignore_permissions=True)


@frappe.whitelist(methods=["POST"])
def dispense_medicine(encounter: str, dispensed_rows: list | str | None = None) -> None:
	doc = load_session_encounter(encounter)
	if doc.custom_workflow_state != "Awaiting Medicine":
		frappe.throw(_("This patient is not awaiting medicine."))

	practitioner = get_nurse_practitioner()
	dispensed_set = set(frappe.parse_json(dispensed_rows) or [])
	rows_by_name = {row.name: row for row in doc.custom_bandhu_prescription}
	for row_name in dispensed_set:
		row = rows_by_name.get(row_name)
		if not row:
			frappe.throw(_("Unknown prescription row."))
		row.dispensed = 1
		row.dispensed_by = practitioner

	doc.custom_workflow_state = "Completed"
	doc.save(ignore_permissions=True)

File: bandhu_app/bandhu_app/page/nurse_form/nurse_form.js
/* global bandhu */

const SESSION_UI_ASSET = "/assets/bandhu_app/js/session_ui.js";

let nurseSession = null;
let encountersByName = {};
let nursePage = null;

async function loadDashboard(page) {
	frappe.dom.freeze();
	let data;
	try {
		const response = await frappe.call({
			method: "bandhu_app.bandhu_app.page.nurse_form.nurse_form.get_session_status",
		});
		data = response.message || {};
	} finally {
		frappe.dom.unfreeze();
	}

	if (!data.has_session) {
		const upcoming = await bandhu.session_ui.get_upcoming_sessions(
			"bandhu_app.bandhu_app.page.nurse_form.nurse_form.get_upcoming_sessions"
		);
		page.main.html(
			'<div class="nurse-dash">' +
				bandhu.session_ui.format_welcome() +
				'<div class="empty-state">' +
				frappe.utils.icon("calendar-off", "xl", "", "", "current-color empty-state-icon") +
				'<span class="empty-state-text">' +
				frappe.utils.escape_html(data.message) +
				"</span></div>" +
				bandhu.session_ui.format_upcoming_sessions(upcoming) +
				"</div>"
		);
		return;
	}

	nurseSession = data;

	if (data.status === "Planned") {
		page.main.html(
			'<div class="nurse-dash">' +
				bandhu.session_ui.format_welcome() +
				bandhu.session_ui.format_session_info(data) +
				'<div class="start-session-bar">' +
				'<button class="btn btn-primary btn-lg nurse-start-session">' +
				frappe.utils.icon("circle-play", "sm", "", "", "current-color") +
				__("Start Session") +
				"</button></div></div>"
		);

		page.main.off("click").on("click", ".nurse-start-session", () => startSession(page));
	} else if (data.status === "In Progress") {
		await loadQueues(page);
	} else if (data.status === "Completed") {
		page.main.html(
			'<div class="nurse-dash">' +
				bandhu.session_ui.format_welcome() +
				bandhu.session_ui.format_session_info(data) +
				'<div class="empty-state">' +
				frappe.utils.icon(
					"circle-check",
					"xl",
					"",
					"",
					"current-color empty-state-icon done"
				) +
				'<span class="empty-state-text">' +
				__("Session completed. Great work!") +
				"</span></div></div>"
		);
	}
}

async function startSession(page) {
	frappe.dom.freeze();
	try {
		await frappe.call({
			method: "bandhu_app.bandhu_app.page.nurse_form.nurse_form.start_session",
			args: { session_name: nurseSession.session_name },
		});
	} finally {
		frappe.dom.unfreeze();
	}

	frappe.show_alert({ message: __("Session started"), indicator: "green" });
	await loadDashboard(page);
}

function endSession(page) {
	frappe.confirm(__("End the current session?"), async () => {
		frappe.dom.freeze();
		try {
			await frappe.call({
				method: "bandhu_app.bandhu_app.page.nurse_form.nurse_form.end_session",
				args: { session_name: nurseSession.session_name },
			});
		} finally {
			frappe.dom.unfreeze();
		}

		frappe.show_alert({ message: __("Session ended"), indicator: "green" });
		await loadDashboard(page);
	});
}

async function loadQueues(page) {
	frappe.dom.freeze();
	const sessionName = nurseSession.session_name;
	let tests, medicines, completed;
	try {
		[tests, medicines, completed] = await Promise.all([
			frappe.call({
				method: "bandhu_app.bandhu_app.page.nurse_form.nurse_form.get_patients_for_tests",
				args: { session_name: sessionName },
			}),
			frappe.call({
				method: "bandhu_app.bandhu_app.page.nurse_form.nurse_form.get_patients_for_medicines",
				args: { session_name: sessionName },
			}),
			frappe.call({
				method: "bandhu_app.bandhu_app.page.nurse_form.nurse_form.get_completed_patients",
				args: { session_name: sessionName },
			}),
		]);
	} finally {
		frappe.dom.unfreeze();
	}

	const testRows = tests.message || [];
	const medicineRows = medicines.message || [];
	const completedRows = completed.message || [];
	encountersByName = Object.fromEntries(
		[...testRows, ...medicineRows, ...completedRows].map((encounter) => [
			encounter.name,
			encounter,
		])
	);

	page.main.html(
		'<div class="nurse-dash">' +
			bandhu.session_ui.format_welcome() +
			bandhu.session_ui.format_session_info(nurseSession) +
			renderEndSessionButton() +
			renderQueueSection(__("Patients for Tests"), testRows, "test") +
			renderQueueSection(__("Patients for Medicines"), medicineRows, "medicine") +
			renderQueueSection(__("Completed Patients"), completedRows, null) +
			"</div>"
	);

	page.main.off("click");

	page.main.on("click", ".nurse-end-session", () => endSession(page));

	page.main.on("click", ".nurse-queue-row", function () {
		frappe.set_route("Form", "Patient Encounter", $(this).data("name"));
	});

	page.main.on("click", ".nurse-action-btn", function (event) {
		event.stopPropagation();
		const encounter = $(this).data("encounter");
		const action = $(this).data("action");
		dispatchNurseAction(page, encounter, action);
	});
}

function dispatchNurseAction(page, encounter, action) {
	switch (action) {
		case "details":
			bandhu.session_ui.open_patient_details_dialog(
				"bandhu_app.bandhu_app.page.nurse_form.nurse_form.get_patient_registration_details",
				encounter,
				encountersByName[encounter] || {}
			);
			break;
		case "enter_results":
			openTestResultsDialog(page, encounter);
			break;
		case "dispense":
			openDispenseDialog(page, encounter);
			break;
		case "vitals":
			openVitalsDialog(page, encounter);
			break;
	}
}

function renderEndSessionButton() {
	return (
		'<div class="end-session-bar">' +
		'<button class="btn btn-danger btn-sm nurse-end-session">' +
		frappe.utils.icon("circle-stop", "sm", "", "", "current-color") +
		__("End Session") +
		"</button></div>"
	);
}

function openTestResultsDialog(page, encounter) {
	const row = encountersByName[encounter];
	if (!row) return;

	const dialog = new frappe.ui.Dialog({
		title: __("Enter Test Results"),
		size: "large",
		fields: [
			{
				fieldtype: "Table",
				fieldname: "results",
				label: __("Tests"),
				cannot_add_rows: true,
				cannot_delete_rows: true,
				in_place_edit: false,
				fields: [
					{
						fieldtype: "Data",
						fieldname: "test_name",
						label: __("Test"),
						in_list_view: 1,
						read_only: 1,
					},
					{
						fieldtype: "Select",
						fieldname: "result_type",
						label: __("Result"),
						options: "\nPositive\nNegative\nValue",
						in_list_view: 1,
					},
					{
						fieldtype: "Data",
						fieldname: "result_value",
						label: __("Value"),
						in_list_view: 1,
					},
					{
						fieldtype: "Small Text",
						fieldname: "notes",
						label: __("Doctor's Notes"),
						read_only: 1,
					},
				],
				data: (row.tests || []).map((test) => ({ ...test })),
			},
		],
		primary_action_label: __("Save Results"),
		primary_action: async (values) => {
			dialog.hide();
			await submitNurseAction(page, "submit_test_results", {
				encounter,
				results: values.results,
			});
		},
	});
	dialog.show();
}

function openDispenseDialog(page, encounter) {
	const row = encountersByName[encounter];
	if (!row) return;

	const dialog = new frappe.ui.Dialog({
		title: __("Dispense Medicine"),
		size: "large",
		fields: [
			{
				fieldtype: "Table",
				fieldname: "prescriptions",
				label: __("Medicines"),
				cannot_add_rows: true,
				cannot_delete_rows: true,
				in_place_edit: false,
				fields: [
					{
						fieldtype: "Data",
						fieldname: "medicines",
						label: __("Medicine"),
						in_list_view: 1,
						read_only: 1,
					},
					{
						fieldtype: "Small Text",
						fieldname: "instructions",
						label: __("Instructions"),
						read_only: 1,
					},
					{
						fieldtype: "Check",
						fieldname: "dispensed",
						label: __("Dispensed"),
						in_list_view: 1,
						default: 1,
					},
				],
				data: (row.prescriptions || []).map((prescription) => ({ ...prescription })),
			},
		],
		primary_action_label: __("Complete"),
		primary_action: async (values) => {
			const dispensedRows = (values.prescriptions || [])
				.filter((prescription) => prescription.dispensed)
				.map((prescription) => prescription.name);
			dialog.hide();
			await submitNurseAction(page, "dispense_medicine", {
				encounter,
				dispensed_rows: dispensedRows,
			});
		},
	});
	dialog.show();
}

function openVitalsDialog(page, encounter) {
	const row = encountersByName[encounter] || {};
	const [bpSystolic, bpDiastolic] = (row.custom_blood_pressure || "").split("/");

	const dialog = new frappe.ui.Dialog({
		title: __("Record Vitals"),
		fields: [
			{
				fieldtype: "Float",
				fieldname: "height_cm",
				label: __("Height (cm)"),
				default: row.custom_height,
			},
			{
				fieldtype: "Float",
				fieldname: "weight_kg",
				label: __("Weight (kg)"),
				default: row.custom_weight,
			},
			{ fieldtype: "Column Break" },
			{
				fieldtype: "Float",
				fieldname: "temperature",
				label: __("Temperature (°F)"),
				default: row.custom_temperature,
			},
			{
				fieldtype: "Int",
				fieldname: "spo2",
				label: __("SpO2 (%)"),
				default: row.custom_spo2,
			},
			{ fieldtype: "Section Break" },
			{
				fieldtype: "Int",
				fieldname: "pulse_rate",
				label: __("Pulse (bpm)"),
				default: row.custom_pulse_rate,
			},
			{
				fieldtype: "Int",
				fieldname: "bp_systolic",
				label: __("BP Systolic"),
				default: bpSystolic || null,
			},
			{ fieldtype: "Column Break" },
			{
				fieldtype: "Int",
				fieldname: "bp_diastolic",
				label: __("BP Diastolic"),
				default: bpDiastolic || null,
			},
		],
		primary_action_label: __("Save Vitals"),
		primary_action: async (values) => {
			dialog.hide();
			await submitNurseAction(page, "record_vitals", {
				encounter,
				height_cm: values.height_cm || null,
				weight_kg: values.weight_kg || null,
				temperature: values.temperature || null,
				pulse_rate: values.pulse_rate || null,
				spo2: values.spo2 || null,
				bp_systolic: values.bp_systolic || null,
				bp_diastolic: values.bp_diastolic || null,
			});
		},
	});
	dialog.show();
}

async function submitNurseAction(page, method, args) {
	frappe.dom.freeze();
	try {
		await frappe.call({
			method: "bandhu_app.bandhu_app.page.nurse_form.nurse_form." + method,
			args,
		});
	} finally {
		frappe.dom.unfreeze();
	}

	frappe.show_alert({ message: __("Saved"), indicator: "green" });
	await bandhu.session_ui.refresh_page(page, loadQueues);
}

function renderQueueActionButtons(encounter, action) {
	const buttons = [
		bandhu.session_ui.format_action_button(
			"nurse-action-btn",
			encounter.name,
			"details",
			__("Details"),
			false
		),
	];
	if (action === "test" || action === "medicine") {
		buttons.push(
			bandhu.session_ui.format_action_button(
				"nurse-action-btn",
				encounter.name,
				"vitals",
				__("Vitals"),
				false
			)
		);
	}
	if (action === "test") {
		buttons.push(
			bandhu.session_ui.format_action_button(
				"nurse-action-btn",
				encounter.name,
				"enter_results",
				__("Enter Results"),
				true
			)
		);
	} else if (action === "medicine") {
		buttons.push(
			bandhu.session_ui.format_action_button(
				"nurse-action-btn",
				encounter.name,
				"dispense",
				__("Dispense"),
				true
			)
		);
	}
	return '<div class="nurse-action-btns">' + buttons.join("") + "</div>";
}

function renderQueueSection(title, encounters, action) {
	const count = '<span class="queue-meta"> (' + encounters.length + ")</span>";

	if (!encounters.length) {
		return (
			'<div class="queue-section">' +
			'<h4 class="queue-head">' +
			frappe.utils.escape_html(title) +
			count +
			"</h4>" +
			'<div class="empty-state">' +
			frappe.utils.icon("inbox", "xl", "", "", "current-color empty-state-icon") +
			'<span class="empty-state-text">' +
			__("No patients in queue.") +
			"</span>" +
			"</div></div>"
		);
	}

	const rows = encounters
		.map(
			(encounter) =>
				'<tr class="nurse-queue-row" data-name="' +
				frappe.utils.escape_html(encounter.name) +
				'">' +
				'<td class="patient-cell">' +
				frappe.utils.escape_html(encounter.patient_name || "") +
				"</td>" +
				'<td class="age-cell">' +
				frappe.utils.escape_html(encounter.patient_age || "") +
				"</td>" +
				'<td class="sex-cell">' +
				frappe.utils.escape_html(encounter.patient_sex || "") +
				"</td>" +
				'<td class="action-cell">' +
				renderQueueActionButtons(encounter, action) +
				"</td>" +
				"</tr>"
		)
		.join("");

	return (
		'<div class="queue-section">' +
		'<h4 class="queue-head">' +
		frappe.utils.escape_html(title) +
		count +
		"</h4>" +
		'<div class="table-wrap">' +
		'<table class="table">' +
		"<thead><tr>" +
		"<th>" +
		__("Patient Name") +
		"</th>" +
		"<th>" +
		__("Age") +
		"</th>" +
		"<th>" +
		__("Sex") +
		"</th>" +
		"<th>" +
		__("Actions") +
		"</th>" +
		"</tr></thead>" +
		"<tbody>" +
		rows +
		"</tbody>" +
		"</table></div></div>"
	);
}

frappe.pages["nurse-form"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: __("Nurse"),
		single_column: true,
	});

	page.set_secondary_action(__("Refresh"), refreshDashboard);
	page.set_primary_action(__("My Schedule"), () => frappe.set_route("my-schedule"), "calendar");

	nursePage = page;
};

async function refreshDashboard() {
	await frappe.require(SESSION_UI_ASSET);
	await bandhu.session_ui.refresh_page(nursePage, loadDashboard);
}

// Desk keeps this page's DOM and module state alive, so returning from a Patient Encounter would
// otherwise show the queues exactly as they were before the encounter was edited. on_page_show
// also fires on the very first show (frappe/public/js/frappe/views/pageview.js:104-107), so it is
// the only loader needed.
frappe.pages["nurse-form"].on_page_show = refreshDashboard;
